File: src/pybennu/pybennu/distributed/provider.py
"""
Provider interface.

This module defines the pybennu Provider base class. It is the interface which
all providers will implement.
"""

import threading
from abc import ABC, abstractmethod

import pybennu.distributed.publisher as publisher
import pybennu.distributed.server as server


class Provider(ABC):
    """
    Provider base class.

    Inheriting providers must implement:

    - query()
    - read(tag)
    - write(tags)
    - periodic_publish()
    """

    def __init__(self, server_endpoint: str, publish_endpoint: str) -> None:
        """Initialize connection environment."""
        self.__publisher = publisher.Publisher(publish_endpoint)

        self.__publish_thread = threading.Thread(target=self.periodic_publish)
        self.__publish_thread.daemon = True

        self.__server = server.Server(server_endpoint)
        self.__server.request_handler = self.__message_handler

    @abstractmethod
    def query(self) -> str:
        """
        Return all current tag names.

        Must return 'ACK=tag1,tag2,...' or 'ERR=<error message>'.
        """
        pass

    @abstractmethod
    def read(self, tag: str) -> str:
        """
        Read the current value of a single tag.

        Must return 'ACK=<value>' or 'ERR=<error message>'.
        """
        pass

    @abstractmethod
    def write(self, tags: dict) -> str:
        """
        Write values to one or more tags.

        Must return 'ACK=<success message>' or 'ERR=<error message>'.
        """
        pass

    @abstractmethod
    def periodic_publish(self) -> None:
        """
        Publish all tags periodically.
        Inheriting providers must implement.
        """
        pass

    def publish(self, msg: str) -> None:
        """Publish message using publisher."""
        self.__publisher.publish(msg)

    def run(self) -> None:
        """Start periodic publish thread and server."""
        self.__publish_thread.start()
        self.__server.run()

    def __message_handler(self, message: str) -> str:
        """
        Message requests should be in the form:
            QUERY=
            READ=<tag name>
            WRITE=<tag name>:<value>[,<tag name>:<value>,...]
            WRITE={tag name:value, tag name:value}

        Note that the last example above uses a JSON blob as the payload
        instead of a plain ol' string.
        """
        if not message:
            return "ERR=Message empty"
        elif '=' not in message:
            return "ERR=Invalid message request"

        split = message.split('=')
        op = split[0]
        payload = split[1]
        print("Received %s request with payload %s" % (op, payload))

        reply = ""
        if op == "QUERY" or op == "query":
            reply += self.query()
        elif op == "READ" or op == "read":
            reply += self.read(payload)
        elif op == "WRITE" or op == "write":
            tags = {}
            for tag in payload.split(','):
                if tag:
                    split = tag.split(':')
                    tags[split[0]] = split[1]
            reply += self.write(tags)
        else:
            reply += "ERR=Unknown command type '%s'", op

        print("Sending reply for payload %s -- %s" % (payload, reply))
        return reply

File: src/bennu/distributed/Provider.cpp
#include "Provider.hpp"

#include <functional>

#include "bennu/distributed/Utils.hpp"

namespace bennu {
namespace distributed {

Provider::Provider(const Endpoint& serverEndpoint, const Endpoint& publishEndpoint)
{
    mPublisher.reset(new Publisher(publishEndpoint));
    mServer.reset(new Server(serverEndpoint));
    mServer->setHandler(std::bind(&Provider::messageHandler, this, std::placeholders::_1));
    printf("Server running on %s\n", serverEndpoint.str.data());
    printf("Publisher running on %s\n", publishEndpoint.str.data());
    fflush(stdout);
}

void Provider::run()
{
    mPublishThread.reset(new std::thread(std::bind(&Provider::periodicPublish, this)));
    mServer->run();
}

// message requests should be in the form:
// "QUERY="
// "READ=<tag name>"
// "WRITE=<tag name>:<value>[,<tag name>:<value>...]"
zmq::message_t Provider::messageHandler(const zmq::message_t& request)
{
    std::string req(request.data<char>());
    std::string opDelim{"="}, op, payload;
    std::string reply;
    try
    {
        auto split = distributed::split(req, opDelim);
        op = split[0];
        payload = split[1];
        printf("Received %s request for payload %s\n", op.data(), payload.data());
    }
    catch (std::exception& e)
    {
        printf("Err: %s", e.what());
    }

    if (op == "QUERY" || op == "query")
    {
        reply += query();
    }
    else if (op == "READ" || op == "read")
    {
        reply += read(payload);
    }
    else if (op == "WRITE" || op == "write")
    {
        std::unordered_map<std::string,std::string> tags;

        std::string pointDelim{","};
        try
        {
            auto split = distributed::split(payload, pointDelim);

            for (auto &point : split)
            {
                std::string valDelim{":"};
                try
                {
                    auto split = distributed::split(point, valDelim);
                    tags.insert({split[0], split[1]});
                }
                catch (std::exception& e)
                {
                    printf("Err: %s", e.what());
                }
            }
        }
        catch(const std::exception& e)
        {
            printf("Err: %s", e.what());
        }

        reply += write(tags);
    }
    else
    {
        reply += "ERR=Unknown command type '" + op + "'";
    }
    printf("Sending reply for payload %s -- %s\n", payload.data(), reply.data());
    zmq::message_t repMsg(reply+'\0'); // must include null byte
    return repMsg;
}

} // namespace distributed
} // namespace bennu
